heroic-core/src/main/java/com/spotify/heroic/cluster/CoreClusterManager.java

```diff
@@ -274,7 +276,33 @@ public void register(LifeCycleRegistry registry) {
             throw new IllegalStateException("Registry not ready");
         }
 
-        return registry.findFromAllShards(OptionalLimit.empty());
+        final List<Pair<Map<String, String>, List<ClusterNode>>> shards =
+            registry.findFromAllShards(OptionalLimit.empty());
+
+        if (expectedShards.isPresent()) {
+            final Set<Map<String, String>> foundShards = new HashSet<>(shards
+                .stream()
+                .map(mapListPair -> mapListPair.getKey())
+                .collect(Collectors.toList()));
+
+            final List<Map<String, String>> shardsWithNoNodes = new ArrayList<>();
+            for (final Map<String, String> expected : expectedShards.get()) {
```

Comment:
With Optionals, prefer unwrapping early to a local variable to make it more obvious what it is.

As an alternative, this block can be rewritten as:

```java
return expectedShards.map(expected -> {
  /* do something with expected */
}).orElse(shards);
```